arch/arm/core/elf.c

```diff
@@ -11,79 +12,333 @@
+	*offset = ((upper & MASK_THM_MOV_IMM4) << SHIFT_THM_MOV_IMM4) |
+		((upper & MASK_THM_MOV_I) << SHIFT_THM_MOV_I) |
+		((lower & MASK_THM_MOV_IMM3) >> SHIFT_THM_MOV_IMM3) | (lower & MASK_THM_MOV_IMM8);
+	*offset = sign_extend(*offset, SHIFT_THM_MOVS_SIGN);
+	*offset += sym_base_addr;
+
+	if (reloc_type == R_ARM_THM_MOVT_PREL || reloc_type == R_ARM_THM_MOVW_PREL_NC) {
+		*offset -= loc;
+	}
+	if (reloc_type == R_ARM_THM_MOVT_ABS || reloc_type == R_ARM_THM_MOVT_PREL) {
+		*offset >>= 16;
+	}
 
-	*((uint16_t *)opaddr) = upper;
-	*(((uint16_t *)opaddr) + 1) = lower;
+	upper = (uint16_t)((upper & (MASK_THM_MOV_11110|MASK_THM_MOV_100100)) |
+			((*offset & (MASK_THM_MOV_IMM4<<SHIFT_THM_MOV_IMM4)) >> SHIFT_THM_MOV_IMM4) |
+			((*offset & (MASK_THM_MOV_I<<SHIFT_THM_MOV_I)) >> SHIFT_THM_MOV_I));
+	lower = (uint16_t)((lower & (MASK_THM_MOV_0|MASK_THM_MOV_RD)) |
+			((*offset & (MASK_THM_MOV_IMM3>>SHIFT_THM_MOV_IMM3)) << SHIFT_THM_MOV_IMM3) |
+			(*offset & MASK_THM_MOV_IMM8));
+	*(uint16_t *)loc = OPCODE2THM16MEM(upper);
+	*(uint16_t *)(loc + 2) = OPCODE2THM16MEM(lower);
 }
 
 /**
- * @brief Architecture specific function for relocating partially linked (static) elf
+ * Decoder functions prefixed by '_thm_' means that they are Thumb instructions specific.
+ * Do NOT mix them with not 'Thumb instructions' in the nelow switch/case: they are not
+ * intended to work together.
  *
- * Elf files contain a series of relocations described in a section. These relocation
- * instructions are architecture specific and each architecture supporting extensions
- * must implement this.
+ * loc = dstsec->sh_addr + rel->r_offset;
  *
- * The relocation codes for arm are well documented
- * https://github.com/ARM-software/abi-aa/blob/main/aaelf32/aaelf32.rst#relocation
+ * loc : addr of opcode
+ * sym_base_addr : addr of symbol if undef else addr of symbol section + sym value
+ * sym_name: symbol name
  */
-void arch_elf_relocate(elf_rela_t *rel, uintptr_t opaddr, uintptr_t opval)
+static int apply_relocate(elf_word reloc_type, uint32_t loc, uint32_t sym_base_addr,
+			const char *sym_name, uintptr_t load_bias)
 {
-	elf_word reloc_type = ELF32_R_TYPE(rel->r_info);
+	int ret = 0;
+	int32_t offset;
+
+	LOG_DBG("%s: %d %x %x %s", __func__, reloc_type, loc, sym_base_addr, sym_name);
 
 	switch (reloc_type) {
+	case R_ARM_NONE:
+		/* no-op */
+		break;
+
 	case R_ARM_ABS32:
-		/* Add the addend stored at opaddr to opval */
-		opval += *((uint32_t *)opaddr);
+		/* fall-through */
+	case R_ARM_TARGET1:
+		*(uint32_t *)loc += sym_base_addr;
+		break;
+
+	case R_ARM_PC24:
+		/* fall-through */
+	case R_ARM_CALL:
+		/* fall-through */
+	case R_ARM_JUMP24:
+		ret = decode_jumps(reloc_type, loc, sym_base_addr, sym_name, &offset);
+		break;
+
+	case R_ARM_V4BX:
+		/* keep Rm and condition bits */
+		*(uint32_t *)loc &= OPCODE2ARMMEM(MASK_V4BX_RM_COND);
+		/* remove the rest */
+		*(uint32_t *)loc |= OPCODE2ARMMEM(MASK_V4BX_NOT_RM_COND);
+		break;
+
+	case R_ARM_PREL31:
```

Comment:
In the 5th commit of the [PR#70171](https://github.com/zephyrproject-rtos/zephyr/pull/70171), there are tests.
I'll add more code to trigger most of the relocations and create a smaller PR when/if the previous ones are merged (otherwise they cannot work)

Reply:
lol, totally agree. I had the same reaction but welcome to [GCC 7](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html) (check for _fallthrough_ or comments)
So either you add a new warning suppressor according to GCC version or you already adapt your code